include/unifex/this.hpp

```diff
@@ -24,95 +24,127 @@ namespace unifex {
 
 struct this_ {};
 
-template <typename ArgType>
-struct replace_this {
-  template <typename Arg, typename T>
+template <typename T>
+inline constexpr bool is_this_v = false;
+template <>
+inline constexpr bool is_this_v<this_> = true;
+template <>
+inline constexpr bool is_this_v<this_&> = true;
+template <>
+inline constexpr bool is_this_v<this_&&> = true;
+template <>
+inline constexpr bool is_this_v<const this_> = true;
+template <>
+inline constexpr bool is_this_v<const this_&> = true;
+template <>
+inline constexpr bool is_this_v<const this_&&> = true;
+
+struct _ignore {
+  template <typename T>
+  /*implicit*/ _ignore(T&&) {}
+};
+
+template <typename>
+struct _replace_this;
+
+template <>
+struct _replace_this<void> {
+  template <typename Arg, typename>
   using apply = Arg;
 
-  template <typename Arg, typename T>
-  static Arg&& get(Arg&& arg, T&) {
+  template <typename Arg>
+  static Arg&& get(Arg&& arg, _ignore) {
     return (Arg &&) arg;
   }
 };
 
 template <>
-struct replace_this<this_> {
-  template <typename Arg, typename T>
+struct _replace_this<this_> {
```

Comment:
Hmm, I wonder if we will also need a specialization for `const this_` at some stage...